homeassistant/components/sensorpush_cloud/api.py

```diff
@@ -0,0 +1,155 @@
+"""Library for interfacing with the SensorPush Cloud API."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import datetime
+from functools import partial, wraps
+import json
+from threading import Lock
+from typing import Any, Concatenate
+
+from sensorpush_api import (
+    AccessTokenRequest,
+    ApiApi,
+    ApiClient,
+    AuthorizeRequest,
+    Configuration,
+    Samples,
+    SamplesRequest,
+    Sensors,
+    SensorsRequest,
+)
+from sensorpush_api.rest import ApiException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.util import dt as dt_util
+
+from .const import ACCESS_TOKEN_EXPIRATION, LOGGER, REQUEST_RETRIES, REQUEST_TIMEOUT
+
+
+def api_call[**_P, _R](
+    func: Callable[Concatenate[SensorPushCloudApi, _P], _R],
+) -> Callable[Concatenate[SensorPushCloudApi, _P], _R]:
+    """Decorate API calls to handle SensorPush Cloud exceptions."""
+
+    @wraps(func)
+    def _api_call(self: SensorPushCloudApi, *args: _P.args, **kwargs: _P.kwargs) -> _R:
+        LOGGER.debug(f"API call to {func} with args={args}, kwargs={kwargs}")
+        retries: int = 0
+        while True:
+            try:
+                result = func(self, *args, **kwargs)
+            except Exception as e:
+                # Force reauthorization if an exception occurs to avoid
+                # authorization failures after temporary outages.
+                self.deadline = dt_util.now()
+
+                # The SensorPush Cloud API suffers from frequent exceptions;
+                # requests are retried before raising an error.
+                if retries < REQUEST_RETRIES:
+                    retries = retries + 1
+                    continue
+
+                LOGGER.debug(f"API call to {func} failed after {retries} retries")
+                if isinstance(e, ApiException):
+                    # API exceptions provide a JSON-encoded message in the
+                    # body; otherwise, fall back to the general behavior.
+                    try:
+                        data = json.loads(e.body)
+                        raise SensorPushCloudError(data["message"]) from e
+                    except Exception:  # noqa: BLE001
+                        pass
+                raise SensorPushCloudError from e
+            else:
+                LOGGER.debug(f"API call to {func} succeeded after {retries} retries")
+                return result
+
+    return _api_call
```

Comment:
This sounds like something thats api specific, should we keep it in the library?

Reply:
Unfortunately the API is a generated Swagger client, which doesn't allow the user to customize generated output beyond the package name etc. I'm not the biggest fan of this approach either, but it seemed the cleanest method to interface with the API without leaking details.

Reply:
Did you generate the openapi?

I always have a huge aversion against those libraries as they always come with deficits and the integration code has to make up for that

Reply:
I did. I was originally going to write a custom API library for it, but after reading the SensorPush developer documentation, they seemed to favor using Swagger for generating API clients instead. There wasn't a huge amount of code to integrate the two, so it seemed like a fair trade-off.

Reply:
Personally I always like the freedom and the efficiency you can add over autogenerating

Reply:
I agree with Joost, that this code should be part of the library